### PR TITLE
feat: instantiate modularized design system 

### PR DESCRIPTION
**Pull request by stephenliang**
[Design Document](https://docs.google.com/document/d/1_gtyjz8skPQLVYLh-4pAy1LgZ5dJziRGxPWz-VsGrUU/edit?tab=t.0#heading=h.g6pya96ifr1h)

This PR instantiates the modularized version of the design system and creates a new monorepo structure for frontend development:

Note: This PR does not build as there is no source code in `frontend/packages/dsco`. However, the purpose is to add components to that package one at a time (or in bulk). The repo does build if you copy components over from `apps/src/componentLibrary` into the `frontend/packages/dsco` directory and applying some patches. (you can revert the "revert dsco components" commit to get this)

1. New top-level directory called `frontend`
2. A devcontainer for `frontend` development
3. A design system storybook in `frontend/apps`

---

1. `devcontainer` scaffolding - to allow devcontainer based development of frontend
2. `apps/design-system-storybook` - Basic Typescript storybook starter plus the addons defined in the design
3. `apps/marketing` - placeholder for CMS
4. `packages/dsco` - tsup starter for compiling sass, ts, tsx into CommonJS and ESM
5. `packages/legacy-css` - Copy of the shared CSS needed for the modularized design system
6. `tsconfig.json` - basic tsconfig (initial one to make the repo work)
7. Two stub components in `packages/dsco/src/` were added to allow the projects to build.

### CODE REVIEW

File: frontend/tsconfig.json
Comment (bencodeorg) [inline frontend/tsconfig.json:1]: Are these defaults derived from somewhere? Or explicit choices?

File: frontend/turbo.json
Comment (bencodeorg) [inline frontend/turbo.json:1]: Same question re: defaults vs. explicit choices

File: frontend/packages/legacy-css/color.scss
Comment (bencodeorg) [inline frontend/packages/legacy-css/color.scss]: Is there a plan for dashboard to consume this package somehow / keep it in sync with what's in `shared/css/color.scss`? Or is the expectation we'll manage it manually?

File: frontend/packages/legacy-css/font.scss
Comment (bencodeorg) [inline frontend/packages/legacy-css/font.scss:132]: Copy over this readme from `shared/css` (or remove this comment)?

File: .devcontainer/frontend/devcontainer.json
Comment (bencodeorg) [inline .devcontainer/frontend/devcontainer.json:3]: Is the expectation that there'd be one devcontainer config for all development within this directory?

File: frontend/README.md
Comment (bencodeorg) [inline frontend/README.md]: Should the description in this readme be updated?

File: frontend/packages/legacy-css/package.json
Comment (bencodeorg) [inline frontend/packages/legacy-css/package.json:11]: Should we throw in basic readmes to match these?

File: frontend/apps/design-system-storybook/.storybook/main.ts
Comment (bencodeorg) [inline frontend/apps/design-system-storybook/.storybook/main.ts:17]: So we retain a webpack dependency for storybook? How do we feel about that?

File: frontend/package.json
Comment (bencodeorg) [inline frontend/package.json]: Should this have a version/etc (this will never be published and is just used to manage the sub-packages, am I right?)

**Review by bencodeorg — COMMENTED** [review]
A few initial clarifications/comments but really cool to see this coming together!

File: frontend/packages/legacy-css/color.scss
Comment (levadadenys) [reply frontend/packages/legacy-css/color.scss]: We won't need this file in `/frontend`. Though in order to that we need to first update the designSystem components to use dsco variables (replace current shared/colors usages with matching pritivieColros.css values)

File: .devcontainer/frontend/devcontainer.json
Comment (stephenliang) [reply .devcontainer/frontend/devcontainer.json:3]: I've been going back and forth on how we can best structure it - I'm thinking one devcontainer for all development in the `frontend` directory may make sense by having the entire toolset available within the turborepo based directory. (Other options were one devcontainer per project which seemed overkill and using the `code-dot-org` one that we did a while back)

The devcontainer here is experimental but the intention is to have it for `frontend` directory for now.

File: frontend/apps/design-system-storybook/.storybook/main.ts
Comment (stephenliang) [reply frontend/apps/design-system-storybook/.storybook/main.ts:17]: Let's chat about this at standup, the initial thinking is that we still need to compile SASS and we can do it in a number of ways, including using the more modern [vite](https://storybook.js.org/docs/get-started/frameworks/react-vite) framework or perhaps the [nextjs](https://storybook.js.org/docs/get-started/frameworks/nextjs) framework. 

Vite would get rid of the dependency on webpack and we'd have to do sass compilation via vite instead. Next.js actually uses webpack under the hood so we'd still need to configure it to compile sass for us. 

The webpack config is pretty lightweight though, it's the styling and `webpackFinal` parts below to allow sass compilation and provide pathing. However we can swap to another framework if that works better for us. In giving consideration to the lightweight webpack config and our continued usage of webpack in Next.js and dashboard I figured using the webpack framework made the most sense. That being said, we can swap to a different framework pretty easily in the future (compared to our current storybook) by swapping out the `framework` line in this file and reconfiguring the configs below.

File: frontend/package.json
Comment (stephenliang) [reply frontend/package.json]: You're right, this should not be here.

File: frontend/packages/legacy-css/color.scss
Comment (stephenliang) [reply frontend/packages/legacy-css/color.scss]: Yes we will want dashboard to consume this and move `shared/css/color.css` to this version. That being said, @levadadenys would you prefer I keep these files out for now? We'll need them eventually since the components use `color.scss` or is the plan that when we move components to the design system we will also do some work to remove references to `color.scss`?

File: frontend/packages/legacy-css/font.scss
Comment (stephenliang) [reply frontend/packages/legacy-css/font.scss:132]: @levadadenys What do you think?

File: frontend/tsconfig.json
Comment (stephenliang) [reply frontend/tsconfig.json:1]: This came from the turborepo starter. Though I think we should do a quick sweep to ensure the options are what we want. I have a [task](https://codedotorg.atlassian.net/browse/ACQ-2793) for creating a common tsconfig on Jira to do that.

File: frontend/turbo.json
Comment (stephenliang) [reply frontend/turbo.json:1]: These are defaults with the exception of line 8 where I added a `dist` output for the design system.

File: frontend/packages/legacy-css/color.scss
Comment (levadadenys) [reply frontend/packages/legacy-css/color.scss]: Sorry, I've written a comment, but didn't posted it yet.

I think we won't need it in /frontend since idea is that all the code-dot-org use DSCO Variables (`common/styles/colors.css`, `common/styles/pirimiteColors.css`)

**Review by levadadenys — COMMENTED** [review]
Super excited to see this becoming reality! 🤩 Thank you!

Left couple comments. I think we can also copy the whole common folder from componentLibrary, except for `apps/src/componentLibrary/common/styles/mixins.scss`.

As for fonts - they'll probably be organized in the different way once I get to refactor and support custom fonts and themes, but should be good for now I guess. We can always delete/rename '@cdo/legacy-css', right?

I also think it's worth settip up a broader readme to `/frontEnd` explaining the content, contributing to it, etc. But that's something can be done in a separate PR as well

File: frontend/packages/legacy-css/color.scss
Comment (stephenliang) [reply frontend/packages/legacy-css/color.scss]: Synced up with @levadadenys and we decided to remove `colors.scss` for now.

**Comment by stephenliang** [comment]
> As for fonts - they'll probably be organized in the different way once I get to refactor and support custom fonts and themes, but should be good for now I guess. We can always delete/rename '@cdo/legacy-css', right?

Yes, the goal is to remove `legacy-css`

> I also think it's worth settip up a broader readme to /frontEnd explaining the content, contributing to it, etc. But that's something can be done in a separate PR as well

I think once we start getting the hang of making components we can document this process in the README, right now lots of processes are still up in the air. There is a task to come back and document everything.

File: frontend/packages/legacy-css/color.scss
Comment (bencodeorg) [reply frontend/packages/legacy-css/color.scss]: Ok -- same question then for other files in this directory like `font.scss`. I don't totally understand how SCSS compilation works in dashboard, but I think these files end up getting dropped into `application.scss` here:

https://github.com/code-dot-org/code-dot-org/blob/c63751fd91bfa3e6d4b77384323aa9019d9fadbe/dashboard/app/assets/stylesheets/application.scss#L34-L37

Would we need to update this pattern if we're getting SCSS from another package? Not sure how much detail to go into here, but basically want to make sure there's a semi-clear path to sharing these files across dashboard and `frontend`.

File: frontend/apps/design-system-storybook/.storybook/main.ts
Comment (bencodeorg) [reply frontend/apps/design-system-storybook/.storybook/main.ts:17]: I'm fine with leaving Webpack here (seems like it'd be unlikely that our config would get particularly complicated for Storybook), but definitely interested to learn more about other options for more complicated applications (I've definitely experienced the complexity of our `apps` Webpack config!)

File: frontend/packages/legacy-css/color.scss
Comment (stephenliang) [reply frontend/packages/legacy-css/color.scss]: Based on the discussion with @levadadenys the plan is to have a copy of font.scss and fontawesome.scss until we develop a long term solution for font loading (so that we don't block component development). So we won't move fonts just yet and this copy is there to allow us to develop components. I'll add a comment to signify this

File: .devcontainer/frontend/devcontainer.json
Comment (bencodeorg) [reply .devcontainer/frontend/devcontainer.json:3]: Can we note somewhere this (and the Dockerfile) are experimental and not the recommended dev workflow (unless they are)?

File: frontend/README.md
Comment (bencodeorg) [inline frontend/README.md]: I think this undersells the amount of stuff in `apps` to someone new to our codebase.

"This directory represents a subset of the frontend code powering Code.org (our marketing site and UI component library). The majority of frontend code powering the teacher and student experience (eg, classroom management, learning platform) still lives in the `apps` directory."

Comment (stephenliang) [reply frontend/README.md]: Great idea, I've updated the README based on your suggestion. Thanks! :)

File: frontend/README.md
Comment (bencodeorg) [inline frontend/README.md:31]: Is there a missing "Install" step here? At least `yarn dev -- --filter=@code-dot-org/design-system` needed an install first

File: frontend/README.md
Comment (bencodeorg) [inline frontend/README.md]: This didn't work until I removed the first --, which aligns with the doc recommendations. Also should be `@code-dot-org/design-system-storybook`

I got it up and running though, pretty cool! 🎉

File: frontend/package.json
Comment (bencodeorg) [inline frontend/package.json:18]: This doesn't actually force using node 22, is that right? We have a .nvmrc that specifies 18.16.0 in `code-dot-org`, how does this interact with that?

File: .devcontainer/frontend/devcontainer.json
Comment (bencodeorg) [reply .devcontainer/frontend/devcontainer.json:3]: Should we just drop the devcontainer config until its more fully baked?

**Review by bencodeorg — APPROVED** [review]
I don't think it's feasible for me to review/understand every bit of config specified here, but the general pieces here make sense to me and align with the design you outlined in the design doc. On a basic level, I was able to get storybook up and running for the design system.

I think it'd be worth getting a second review from someone in the broader engineering team (even if you merge this as-is in the meantime to continue progress), as these are decisions that will affect at least acquisition products as well as whoever ends up responsible for maintaining the shared `frontend` config and the design system package.

I also think linking to the design doc in your PR description would be helpful for future readers to understand the broader context here (bonus points if the description itself included some of the main design decision, as Brendan once told me, "Jira/Slack/Google Docs/etc. are temporary, but PR descriptions are forever".

All of that said, Thank you for your leadership here!

**Comment by stephenliang** [comment]
> I think it'd be worth getting a second review from someone in the broader engineering team (even if you merge this as-is in the meantime to continue progress), as these are decisions that will affect at least acquisition products as well as whoever ends up responsible for maintaining the shared `frontend` config and the design system package.

Let me send a request for comment to `#apps-dev` but in the meantime I'll start a `cms` branch which tracks `staging`. We can work off of `cms` so that it's not blocking and we can add updates as a follow up PR there.

File: frontend/package.json
Comment (stephenliang) [reply frontend/package.json:18]: It should force node 22 (yarn will fail), but you're right that I should add a `.nvmrc` that specifies node 22.  

I'll create a new task to add that file.

Comment (bencodeorg) [reply frontend/package.json:18]: I successfully `yarn install`-ed and don't have Node 22 installed.

File: .devcontainer/frontend/devcontainer.json
Comment (stephenliang) [reply .devcontainer/frontend/devcontainer.json:3]: I'd recommend we keep it for now since it needs to be there in order for development to work in the first place. That being said, I think we'll  use the `cms` branch so it won't land in `staging` just yet until we solicit more feedback from other engineering teams.

File: frontend/package.json
Comment (stephenliang) [reply frontend/package.json:18]: Interesting, looks like yarn 2+ [removed the check](https://github.com/yarnpkg/berry/issues/1177) but yarn 1 (classic) still has it. I recommend we keep the check in case yarn decides to add it in the future and it helps other tools in the future.

```
yarn install v1.22.22
warning Missing version in workspace at "/Users/admin/git-workspaces/code-dot-org/frontend/apps/design-system-storybook", ignoring.
[1/5] 🔍  Validating package.json...
warning Missing version in workspace at "/Users/admin/git-workspaces/code-dot-org/frontend/apps/design-system-storybook", ignoring.
error @code-dot-org/frontend@: The engine "node" is incompatible with this module. Expected version ">=22". Got "18.20.4"
error Found incompatible module.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

File: frontend/README.md
Comment (stephenliang) [reply frontend/README.md]: You're right - I had renamed it and forgot to update the doc. I also changed it to use `yarn workspace`, so the command should be:

```
yarn workspace @code-dot-org/design-system-storybook dev
```

File: frontend/README.md
Comment (stephenliang) [reply frontend/README.md:31]: Yes I have added this

File: .devcontainer/frontend/devcontainer.json
Comment (bencodeorg) [reply .devcontainer/frontend/devcontainer.json:3]: > "In order for development to work in the first place"

Is the implication that all development must take place in a devcontainer?

**Pull request merged**
